File: tests/test_element.cxx
/**
 *
 * Unit tests for Element class
 *
 * (c) 2019
 * Kai Krings
 *
 * @file test_element.cxx
 * @date 2019-08-26
 * @author Kai Krings <kai.krings@posteo.de>
 *
 */
#include <stdexcept>

#include <gtest/gtest.h>

#include "seawater/element.h"

/**
 * Test the static member function GetMinCharge.
 */
TEST(ElementTestCase, TestMinCharge)
{
    EXPECT_EQ(seawater::Element::GetMinCharge(), 1);
}

/**
 * Test the static member function GetMinWeight.
 */
TEST(ElementTestCase, TestMinWeight)
{
    EXPECT_DOUBLE_EQ(seawater::Element::GetMinWeight(), 1.);
}

/**
 * Test the constant member variable symbol_.
 */
TEST(ElementTestCase, TestSymbol)
{
    seawater::Element hydrogen{"H", 1, 1.};
    EXPECT_EQ(hydrogen.GetSymbol(), "H");
}

/**
 * Test the constant member variable charge_.
 */
TEST(ElementTestCase, TestCharge)
{
    seawater::Element hydrogen{"H", 1, 1.};
    EXPECT_EQ(hydrogen.GetCharge(), 1);
}

/**
 * Test expected exceptions from initialization.
 */
TEST(ElementTestCase, TestChargeException)
{
    EXPECT_THROW(seawater::Element("H", 0, 1.), std::invalid_argument);
}

/**
 * Test the constant member variable weight_.
 */
TEST(ElementTestCase, TestWeight)
{
    seawater::Element hydrogen{"H", 1, 1.};
    EXPECT_DOUBLE_EQ(hydrogen.GetWeight(), 1.);
}

/**
 * Test expected exceptions from initialization.
 */
TEST(ElementTestCase, TestWeightException)
{
    EXPECT_THROW(seawater::Element("H", 1, 0.), std::invalid_argument);
}

/**
 * Test the less than operator.
 */
TEST(ElementTestCase, TestLTOperator)
{
    seawater::Element hydrogen{"H", 1, 1.};
    seawater::Element oxygen{"O", 8, 16.};

    EXPECT_LT(hydrogen, oxygen)
        << "Hydrogen is expected to be smaller than oxygen "
        << "in terms of charge.";
}
/**
 * Test the greater than operator.
 */
TEST(ElementTestCase, TestGTOperator)
{
    seawater::Element hydrogen{"H", 1, 1.};
    seawater::Element oxygen{"O", 8, 16.};

    EXPECT_GT(oxygen, hydrogen)
        << "Oxygen is expected to be greater than hydrogen "
        << "in terms of charge.";
}

File: src/tests/pyseawater/test_element.py
# -*- coding: utf-8 -*-

"""Unit tests for `pyseawater.Element`

"""
import unittest

import pyseawater


class ElementTestCase(unittest.TestCase):
    """Test cases for `pyseawater.Element`

    """
    def test_min_charge(self):
        """Test static property `min_charge`.

        """
        self.assertEqual(
            pyseawater.Element.min_charge, 1,
            "The static property `min_charge` is not equal ``1``.")

    def test_min_weight(self):
        """Test static property `min_weight`.

        """
        self.assertEqual(
            pyseawater.Element.min_weight, 1.,
            "The static property `min_weight` is not equal ``1.``.")

    def test_symbol(self):
        """Test property `symbol`.

        """
        hydrogen = pyseawater.Element('H', 1, 1.)

        self.assertEqual(
            hydrogen.symbol, 'H',
            "The property `symbol` is not equal the expected value ``'H'``.")

    def test_charge(self):
        """Test property `charge`.

        """
        hydrogen = pyseawater.Element('H', 1, 1.)

        self.assertEqual(
            hydrogen.charge, 1,
            "The property `charge` is not equal the expected value ``1``.")

    def test_charge_exception(self):
        """Test expected exceptions from initialization.

        """
        message = "A `charge` of ``0`` did not raise an exception."

        with self.assertRaises(ValueError, msg=message):
            pyseawater.Element('H', 0, 1.)

    def test_weight(self):
        """Test property `weight`.

        """
        hydrogen = pyseawater.Element('H', 1, 1.)

        self.assertEqual(
            hydrogen.weight, 1.,
            "The property `weight` is not equal the expected value ``1.``.")

    def test_weight_exceptions(self):
        """Test expected exceptions from initialization.

        """
        message = "A `weight` of ``0.`` did not raise an exception."

        with self.assertRaises(ValueError, msg=message):
            pyseawater.Element('H', 1, 0.)

    def test_lt_operator(self):
        """Test less than operator.

        """
        hydrogen = pyseawater.Element('H', 1, 1.)
        oxygen = pyseawater.Element('O', 8, 16.)

        message = (
            "Hydrogen is expected to be smaller than oxygen in terms of "
            "charge."
            )

        self.assertLess(hydrogen, oxygen, message)

    def test_gt_operator(self):
        """Test greater than operator.

        """
        hydrogen = pyseawater.Element('H', 1, 1.)
        oxygen = pyseawater.Element('O', 8, 16.)

        message = (
            "Oxygen is expected to be larger than hydrogen in terms of "
            "charge."
            )

        self.assertGreater(oxygen, hydrogen, message)
